dbms/src/Storages/MergeTree/MergeTreeReaderCompact.cpp

```diff
@@ -0,0 +1,272 @@
+
+                size_t read_rows_in_column = column->size() - column_size_before_reading;
+
+                if (read_rows_in_column < rows_to_read)
+                    throw Exception("Cannot read all data in MergeTreeReaderCompact. Rows read: " + toString(read_rows_in_column) +
+                        ". Rows expected: " + toString(rows_to_read) + ".", ErrorCodes::CANNOT_READ_ALL_DATA);
+            }
+            catch (Exception & e)
+            {
+                /// Better diagnostics.
+                e.addMessage("(while reading column " + name + ")");
+                throw;
+            }
+        }
+
+        ++from_mark;
+        read_rows += rows_to_read;
+    }
+
+    for (size_t i = 0; i < num_columns; ++i)
+    {
+        auto & column = mutable_columns[i];
+        if (column && column->size())
+            res_columns[i] = std::move(column);
+        else
+            res_columns[i] = nullptr;
+    }
+
+    next_mark = from_mark;
+
+    return read_rows;
+}
+
+MergeTreeReaderCompact::ColumnPosition MergeTreeReaderCompact::findColumnForOffsets(const String & column_name)
+{
+    String table_name = Nested::extractTableName(column_name);
+    for (const auto & part_column : data_part->getColumns())
+    {
+        if (typeid_cast<const DataTypeArray *>(part_column.type.get()))
+        {
+            auto position = data_part->getColumnPosition(part_column.name);
+            if (position && Nested::extractTableName(part_column.name) == table_name)
+                return position;
+        }
+    }
+
+    return {};
+}
+
+
+void MergeTreeReaderCompact::readData(
+    const String & name, IColumn & column, const IDataType & type,
+    size_t from_mark, size_t column_position, size_t rows_to_read, bool only_offsets)
+{
+    if (!isContinuousReading(from_mark, column_position))
+        seekToMark(from_mark, column_position);
+
+    auto buffer_getter = [&](const IDataType::SubstreamPath & substream_path) -> ReadBuffer *
+    {
+        if (only_offsets && (substream_path.size() != 1 || substream_path[0].type != IDataType::Substream::ArraySizes))
+            return nullptr;
+
+        return data_buffer;
+    };
+
+    IDataType::DeserializeBinaryBulkSettings deserialize_settings;
+    deserialize_settings.getter = buffer_getter;
+    deserialize_settings.avg_value_size_hint = avg_value_size_hints[name];
+    deserialize_settings.position_independent_encoding = true;
+
+    IDataType::DeserializeBinaryBulkStatePtr state;
+    type.deserializeBinaryBulkStatePrefix(deserialize_settings, state);
+    type.deserializeBinaryBulkWithMultipleStreams(column, rows_to_read, deserialize_settings, state);
+
+    /// The buffer is left in inconsistent state after reading single offsets
+    if (only_offsets)
+        last_read_granule.reset();
+    else
+        last_read_granule.emplace(from_mark, column_position);
+}
+
+
+void MergeTreeReaderCompact::initMarksLoader()
+{
+    if (marks_loader.initialized())
```

Comment:
Seems like it became initialized when the constructor is called. Maybe it can be wrapped into `std::optional`?